src/Conversion/ONNXToTOSA/Tensor/PaddingOp.cpp

```diff
@@ -0,0 +1,169 @@
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "llvm/ADT/None.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Casting.h"
+
+#ifdef _WIN32
+#include <io.h>
+#endif
+
+using namespace mlir;
+
+//  ONNX Pad operation
+//    “Given a tensor containing the data to be padded (data),
+//    a tensor containing the number of start and end pad values for
+//    axis (pads), (optionally) a mode, and (optionally) constant_value,
+//    ” “a padded tensor (output) is generated.
+//
+//  Attributes:
+//    mode	::mlir::StringAttr	string attribute
+//
+//  Operands:
+//  data	  tensor of 8-bit/16-bit/32-bit/64-bit unsigned integer values
+//  or 	  tensor of 8-bit/16-bit/32-bit/64-bit signless integer values or
+//  tensor of bfloat16 type values or tensor of 16-bit/32-bit/64-bit 	  float
+//  values or tensor of string type values or tensor of 1-bit 	  signless
+//  integer values or tensor of complex type with 	  32-bit/64-bit float
+//  elements values or memref of any type values.
+//
+//  pads   tensor of 64-bit signless integer values or memref of
+//  	  any type values.
+//
+//  constant_value
+//  	  tensor of 8-bit/16-bit/32-bit/64-bit unsigned integer values or
+//         tensor of 8-bit/16-bit/32-bit/64-bit signless integer values or
+//         tensor of bfloat16 type values or tensor of 16-bit/32-bit/64-bit
+//         float values or tensor of string type values or tensor of 1-bit
+//         signless integer values or tensor of complex type with
+//         32-bit/64-bit float elements values or memref of any type values
+//         or none type.
+//
+// Results:
+//  output
+//         tensor of 8-bit/16-bit/32-bit/64-bit unsigned integer values or
+//         tensor of 8-bit/16-bit/32-bit/64-bit signless integer values or
+//         tensor of bfloat16 type values or tensor of 16-bit/32-bit/64-bit
+//         float values or tensor of string type values or tensor of 1-bit
+//         signless integer values or tensor of complex type with
+//         32-bit/64-bit float elements values or memref of any type values
+//         or none type.
+//
+
+namespace onnx_mlir {
+
+class ONNXPadOpLoweringToTOSA : public OpConversionPattern<ONNXPadOp> {
+public:
+  using OpConversionPattern<ONNXPadOp>::OpConversionPattern;
+  using OpAdaptor = typename ONNXPadOp::Adaptor;
+  LogicalResult matchAndRewrite(ONNXPadOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const final {
+
+    Location loc = op.getLoc();
+
+    Value data = adaptor.data();
+    Value pads = adaptor.pads();
+    Value constValue = adaptor.constant_value();
+
+    if (!(adaptor.mode() == "constant")) {
+      return rewriter.notifyMatchFailure(
+          op, "Only 'constant' mode is supported");
+    }
```

Comment:
`pads` and `constant_value` need not be constant - see for example https://github.com/ljfitz/onnx-mlir/blob/4900aca663fb029f4a628f7a3bd6200d0141bbbb/test/mlir/onnx/onnx_lowering_with_canonicalize.mlir#L2029:L2031. You might also want to reject them here if `tosa` cannot easily support them.

Reply:
You are right, I added a test, too. This is tricky because one could also probably do a lot of reshaping to make this work. Creating the right pad tensor will be very hard then. I am curious what the upstream people think